### ISSUE TITLE
Browser support (why not?)

### ISSUE THREAD

**Issue by mindplay-dk** [chalk/chalk#535]
Please (re)consider adding support for the browser.

I mean, why not? Chrome supports ANSI escape codes since version 60-something, so you don't need an alternative approach using styles or anything - the package probably just needs to be made compatible with the browser somehow?

If you [try it in the browser](https://codesandbox.io/s/hopeful-night-n6oumh?file=/src/index.js) currently, you get:

    Could not find module in path: 'chalk/source/#ansi-styles' relative to '/node_modules/chalk/source/index.js'

I also tried `kleur`, `colorette` and `picocolors` - these can all *load* in the browser, though none of them actually *work*; they simply don't generate any color codes, which may have something to do with how color libraries detect support for colors on the console.

After trying numerous competing color libraries, I eventually discovered `ansi-colors-es6` - the only library I've found that [does work](https://codesandbox.io/s/headless-platform-8idzcq?file=/src/index.js).

So 1 out of 5 libraries worked, heh - and it's the least popular of all of these libraries, go figure.

Perhaps this is not something *most* people care about?

But the [first issue ever opened](https://github.com/chalk/chalk/issues/1) asked for it. It's been [asked for](https://github.com/chalk/chalk/issues/10), [asked for](https://github.com/chalk/chalk/issues/50), [misunderstood](https://github.com/chalk/chalk/issues/143), [asked for](https://github.com/chalk/chalk/issues/284), [asked for](https://github.com/chalk/chalk/issues/424) and [misunderstood](https://github.com/chalk/chalk/issues/503) yet again, so that's gotta make you think, right? 😏

I don't see anything in `ansi-colors-es6` to suggest they've had to make any substantial concessions in terms of either bundle size or performance - it remains comparatively fast and small. (of course, it claims to be the fastest, which, oddly, most color libraries do... but either way, as long as it's not *slow* - if you're competing on marginal percentages - who cares.)

Bottom line, it's been solved, so there's a source of reference for the solution, and it doesn't look bad, slow, complex or ugly or anything. It appears to be a drop-in replacement for most of these color libraries, which are all pretty similar... but wouldn't it be nice if people could just install the most popular color library and have it actually work as expected?

Why not? 🙂

**Comment by sindresorhus** [comment]
Looks like CodeSandbox does not support [subpath imports](https://nodejs.org/api/packages.html#subpath-imports). I would open an issue on their issue tracker instead: https://github.com/codesandbox

**Comment by mindplay-dk** [comment]
I think this a node-only feature? I had never heard of it.

The package doesn't load in stackblitz or codepen either, and doesn't produce a working ES6 module on skypack... so it seems whatever this feature is, it isn't very friendly to browsers.

**Comment by NickKelly1** [comment]
You're better off downgrading to v4 and finding an alternative to this package. There isn't support or backwards compatibility here.

**Comment by Jakobeha** [comment]
Hello,

I made a fork of chalk with node + browser support for internal projects I'm working on. See https://github.com/Jakobeha/chalk-cross. There are also a couple other features (exposes open and close escape codes)

It adds 16m escape codes to anything on the browser. If that doesn't work, you will have to change src/shims/chalk-common/index.js#getModelAnsi to handle 'custom' differently.

**Comment by mindplay-dk** [comment]
I eventually went with `ansi-colors` - it works in the browser.

**Comment by dimaslanjaka** [comment]
CSS be like: What about me ?

**Comment by sindresorhus** [comment]
> I think this a node-only feature? I had never heard of it.

It's a Node.js feature, yes. But it's being adopted by build tools too.

> so it seems whatever this feature is, it isn't very friendly to browsers.

It's not a browser problem, but rather these services not using the latest version of the build tools.

**Comment by sindresorhus** [comment]
We cannot really remove this feature as it's required to be able to correctly provide a browser version of a sub-module we use in Chalk.

https://github.com/chalk/chalk/blob/b17d862944f9efea1ee12aaf80fe19ed5063a34d/package.json#L14-L15

**Comment by mindplay-dk** [comment]
I'm afraid I don't follow. You actually have color support detection code specifically for the browser - but if the module can't be loaded in the browser in the first place, what is this for?

> It's not a browser problem, but rather these services not using the latest version of the build tools.

For the record, it didn't work with current versions of rollup and standard plugins.

I mean, I don't fully understand what's missing or whether it's on a roadmap for common bundlers somewhere in the future. I just picked a package, and it doesn't really seem to work anywhere the browser is involved.

(Maybe no one else cares? I'm building a reporter that will run both in the browser and on the CLI, so this was important to me. Either way, `ansi-colors` works fine and does the same thing, so... case closed as far as I care.)

**Comment by Qix-** [comment]
Perhaps the key point in Sindre's comment was missed:

The problem you're running into has nothing to do with browsers. It has to do with your bundlers not supporting this feature of path resolution. Once whatever bundler you're using does (or you update) this will begin to work in the browser.
